[hiivan] iP

hiivan opened the pull request:
# Huyang Chatbot 🤖

> "Your mind is for having ideas, not holding them." – David Allen ([source](https://gettingthingsdone.com/))

**Huyang** frees your mind of having to remember things you need to do. It's,

- *text-based*
- **easy to learn**
- ***FAST SUPER FAST*** to use

All you need to do is,

1. download it from [here](https://github.com/yourusername/huyang/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

### Features:

- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [x] `delete`, `find`, `list` functions

To create a JAR file, run `./gradlew shadowJar` in your terminal.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

hiivan committed:
Rename Duke to Huyang

hiivan committed:
Level-0

hiivan committed:
added echo method

hiivan committed:
store tasks and display

hiivan committed:
included Task.java

hiivan committed:
mark/unmark feature

hiivan committed:
Updated Task.java

hiivan committed:
add support for ToDos, Events, Deadlines

hiivan committed:
Updated text-ui-test

hiivan committed:
Updated bugs in main/java

hiivan committed:
Added robust error handling

hiivan committed:
Updated robust error handling

hiivan committed:
Added Delete feature

hiivan committed:
Added enums and improved automated testing

hiivan committed:
Level 7: Save

hiivan committed:
Merge branch 'branch-Level-7'
Level-7 merge required

hiivan committed:
Level 8: Date and Time

hiivan committed:
Merge branch 'branch-Level-8'
Level-8 merge required

hiivan committed:
Made it more OOP

hiivan committed:
Merge branch 'branch-A-MoreOOP'
i hate oop

hiivan committed:
Organize into a single package

hiivan committed:
Add JUnit testing

JUnit tests are absent, limiting our ability to
ensure code quality.

To ensure future changes do not break existing
functionality, it is crucial to introduce unit
testing.

Add JUnit testing framework to the project to
cover critical functions with unit tests.

The JUnit framework is chosen for its wide
adoption, ease of use, and integration
capabilities with existing Java projects.

This commit does not cover all possible test
cases but establishes a foundation for future
test development and continuous integration
processes.

hiivan committed:
Package the App as a JAR file huyang.jar

hiivan committed:
Implement findTask()

hiivan committed:
Add JavaDoc comments for documentation purposes

This commit enhances JavaDocs for core classes
and public methods, focusing on clear
explanations, parameters, and return values to
improve codebase accessibility.

By documenting critical components, we aim to
facilitate easier onboarding for new
contributors and support efficient code
maintenance.

hiivan committed:
Update JavaDoc for find function

alfaloo commented on build.gradle:
```diff
@@ -0,0 +1,42 @@
+
+run{
+    standardInput = System.in
+}
```
Looks good I approve

alfaloo commented on gradlew:
```diff
@@ -0,0 +1,248 @@
+        tr '\n' ' '
+    )" '"$@"'
+
+exec "$JAVACMD" "$@"
```
I like this piece of code

alfaloo commented on gradlew.bat:
```diff
@@ -0,0 +1,92 @@
+:mainEnd
+if "%OS%"=="Windows_NT" endlocal
+
+:omega
```
can these code by simplified?

alfaloo commented on src/main/java/huyang/Event.java:
```diff
@@ -0,0 +1,96 @@
+        return getTypeIcon() + super.getStatusIcon() + " " + taskName +
+                " (from: " + start.format(formatter) + " to: " + end.format(formatter) + ")";
+    }
+}
```
can you improve on these?

alfaloo reviewed (commented):
I like the changes

hiivan committed:
Update README.md

hiivan committed:
Update coding violations found by checkstyle

hiivan committed:
Initial JavaFX prototype with some major code refactoring

hiivan committed:
Rebuild the scene using FXML

hiivan committed:
Add greeting message to GUI and update JavaDocs

donwong2308 commented on src/main/java/huyang/Huyang.java:
```diff
@@ -0,0 +1,157 @@
+            try {
+                switch (command) {
+                case LIST:
+                    ui.print(ui.getTasksMessage(tasks.getTasks()));
+                    break;
+                case MARK:
+                    Task markedTask = tasks.markOrUnmarkTask(input, true);
+                    ui.print(ui.getMarkOrUnmarkMessage(markedTask, true));
+                    storage.saveTasks(tasks.getTasks());
+                    break;
+                case UNMARK:
+                    Task unmarkedTask = tasks.markOrUnmarkTask(input, false);
+                    ui.print(ui.getMarkOrUnmarkMessage(unmarkedTask, false));
+                    storage.saveTasks(tasks.getTasks());
+                    break;
+                case TODO:
+                case DEADLINE:
+                case EVENT:
+                    Task addedTask = tasks.addTask(input, command);
+                    ui.print(ui.getAddTaskMessage(addedTask, tasks.getSize()));
+                    storage.saveTasks(tasks.getTasks());
+                    break;
+                case DELETE:
+                    Task deletedTask = tasks.deleteTask(input);
+                    ui.print(ui.getDeleteTaskMessage(deletedTask, tasks.getSize()));
+                    storage.saveTasks(tasks.getTasks());
+                    break;
+                case FIND:
+                    String keyword = input.substring(5).trim().toLowerCase();
+                    ArrayList<Task> foundTasks = tasks.findTasks(keyword);
+                    ui.print(ui.getFoundTasksMessage(foundTasks));
+                    break;
+                case BYE:
+                    isExit = true;
+                    break;
+                case UNKNOWN:
+                default:
+                    ui.print(ui.getUnknownCommandMessage());
+                    break;
+                }
```
I like the use of switch case here to make the code more readable and easier to debug 👍

hiivan replied:
Thank you!

donwong2308 commented on src/main/java/huyang/TaskFactory.java:
```diff
@@ -0,0 +1,68 @@
+import java.time.format.DateTimeParseException;
+
+public class TaskFactory {
+    private static final String DATE_PATTERN = "yyyy-MM-dd HHmm";
```
Would it be possible to add more date patterns or make the time optional?

hiivan replied:
Yes, you can always change the date pattern by editing what the static final variable is. To make it optional, simply comment this out and replace DATE_PATTERN with whatever date pattern you want.

donwong2308 commented on src/main/java/huyang/Deadline.java:
```diff
@@ -0,0 +1,77 @@
+ * Represents a deadline task with a specified deadline time.
+ */
+public class Deadline extends Task {
+    protected LocalDateTime by;
```
Perhaps the variable name can be deadineDate instead for clarity?

hiivan replied:
Yes, that should make my code clearer! Thank you.

donwong2308 reviewed (commented):
Overall, everything looks good to me. Very readable code and easy to understand 😃

shawnnlimm commented on src/main/java/huyang/Event.java:
```diff
@@ -0,0 +1,97 @@
+        return end;
+    }
+
+    private String getTypeIcon() {
+        return "[E]";
+    }
```
Perhaps this could just be a private final attribute instead of a method?

hiivan replied:
Since this is a private attribute already, is there a point in making it final?

shawnnlimm commented on src/main/java/huyang/Huyang.java:
```diff
@@ -0,0 +1,157 @@
+                Task unmarkedTask = tasks.markOrUnmarkTask(input, false);
+                storage.saveTasks(tasks.getTasks());
+                return ui.getMarkOrUnmarkMessage(unmarkedTask, false);
+            case TODO:
+            case DEADLINE:
```
You might have missed out the implementation here? Same in line 76

hiivan replied:
The `switch` data structure allows me to group several conditions in the same case statement using the mechanism of _fall through_.

shawnnlimm commented on src/main/java/huyang/Parser.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * Enumeration representing the possible command types.
+     */
+    public enum CommandType {
+        LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND, BYE, UNKNOWN;
+    }
```
I like how you used enums here for the Commands!

hiivan replied:
Thank you!

shawnnlimm commented on src/main/java/huyang/Task.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * An abstract class representing a task with a name and completion status.
+ */
+public abstract class Task {
```
I like how you used an abstract class to model the Task class!

shawnnlimm commented on src/main/java/huyang/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+     * @return The modified task.
+     * @throws TaskException If there is an issue marking or unmarking the task.
+     */
+    public Task markOrUnmarkTask(String input, boolean isDone) throws TaskException {
+        Task task = getTaskByNumber(input);
+        task.setDone(isDone);
+        return task;
+    }
```
Might it be better to split this method into 2 separate methods? For eg. markTask and unmarkTask

hiivan replied:
I felt that this way was more elegant as the logic of marking or unmarking is contained within the `markOrUnmarkTask()` method.

shawnnlimm reviewed (commented):
Overall, your code is easy to read and there are no major style violations. Your code also follows the coding standards well. Good job!

hiivan committed:
Change storage filepath

hiivan committed:
Implement assertions for data validation

Added assertions to check data integrity.
Targets crucial methods and aims for early
bug detection and code clarity.

Note: Assertions are enabled during
development phases only.

hiivan committed:
Improve code quality

Changed some of the code to comply with
the NUS coding standard.

hiivan committed:
Optimize findTasks with Java Streams

Refactored the findTasks method to utilize Java
Streams, enhancing readability and efficiency.

The method now filters tasks by keyword in a
more streamlined manner, ensuring
case-insensitive matching for greater
reliability.

This approach not only simplifies the code but
also aligns with modern Java practices,
promoting maintainability and scalability.

hiivan committed:
Merge pull request #1 from hiivan/branch-A-Assertions

Implement assertions for data validation

hiivan committed:
Merge pull request #2 from hiivan/branch-A-CodeQuality

Improve code quality

hiivan committed:
Merge pull request #3 from hiivan/branch-A-Streams

Optimize findTasks with Java Streams

hiivan committed:
Personalize messages to sound like Huyang

hiivan committed:
Improve GUI

hiivan committed:
Add and refine JUnit tests

hiivan committed:
Update user guide and add UI.png

hiivan committed:
Update README.md

hiivan committed:
Change Ui class methods to all static

hiivan committed:
Add directory and file creation checks to Storage constructor

hiivan committed:
Update User Guide to include instructions for various commands
